[SERVER-1469] server 3.3 - adding docs for nomad-autoscaler on server (#6066)
circleci/circleci-docs: jekyll/_cci2/server-3-install-build-services.adoc

* *Nomad Server Certificate Authority (CA) Certificate (required)* - 
Provided in the output from `terraform apply`

==== Nomad Autoscaler (Optional)

> Nomad provides a utility called to automatically scale up or down your nomad clients provided your clients are managed a cloud providers autoscaling resource. With Nomad Autoscaler, you only need to provide permission for the utility to manage your autoscaling resource and where it is located. You can enable this resource via KOTS, which will deploy the nomad autoscaler service along with your nomad servers. Below we will go through how to setup nomad autoscaler for your provider.

Reviewer: ```suggestion
Nomad provides a utility to automatically scale up or down your nomad clients provided your clients are managed by a cloud provider's autoscaling resource. With Nomad Autoscaler, you only need to provide permission for the utility to manage your autoscaling resource and where it is located. You can enable this resource via KOTS, which will deploy the Nomad Autoscaler service along with your Nomad servers. Below we will go through how to set up Nomad Autoscaler for your provider.
```

> If you do not require this service then feel free to jump tp the *Save config* button to update your installation and re-deploy server.

Reviewer: ```suggestion
If you do not require this service then feel free to jump to the *Save config* button to update your installation and re-deploy server.
```

===== AWS
. Create an IAM user/role and policy for Nomad Autoscaler. You may take one of the following approaches:

>   * Our link:https://github.com/CircleCI-Public/server-terraform/tree/main/nomad-aws[nomad module] creates an IAM user and outputs the keys if you set variable `nomad_auto_scaler = true`. You may reference the example in the link for more details. If you've already created the clients, you can update the variable and run `terraform apply`. The created user's access key and secret will be available in Terraform's output.

Reviewer 2: It feels like we should talk about this step earlier so if people know they want to autoscale they can do this at the point of  deploying their nomad clients for the first time.

Also, this feels like something customers want to do after the initial install so I will make a server 3 specific guide to nomad to include this info (up to now we've just been using the nomad guide from v2 - https://circleci.com/docs/2.0/nomad/) - I'll do this in a separate PR once we've merged this into the 3.3 branch though. Let me know if that doesn't sound right.

Author: I wasnt sure myself. My thinking though is for first install, they may want to focus on getting things running and can return to autoscaling. its an optional step and does no harm to do this later. 
However I dont have a strong opinion here and am guided by the your experience.

Reviewer 2: I put a note in above to flag this section up. Let's leave as is for now, can always rearrange later.

>   * Our link:https://github.com/CircleCI-Public/server-terraform/tree/main/nomad-aws[nomad module] creates an IAM user and outputs the keys if you set variable `nomad_auto_scaler = true`. You may reference the example in the link for more details. If you've already created the clients, you can simply update the variable and run `terraform apply`. The created user's access key and secret will be available in terraform's output.

Reviewer: ```suggestion
  * Our link:https://github.com/CircleCI-Public/server-terraform/tree/main/nomad-aws[nomad module] creates an IAM user and outputs the keys if you set variable `nomad_auto_scaler = true`. You may reference the example in the link for more details. If you've already created the clients, you can simply update the variable and run `terraform apply`. The created user's access key and secret will be available in Terraform's output.
```

  * You may also create a nomad IAM user manually. The user will need an IAM policy with the following rules attached:
[source, json]
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "autoscaling:CreateOrUpdateTags",
                "autoscaling:UpdateAutoScalingGroup",
                "autoscaling:TerminateInstanceInAutoScalingGroup"
            ],
            "Resource": "<<Your Autoscaling Group ARN>>"
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": [
                "autoscaling:DescribeScalingActivities",
                "autoscaling:DescribeAutoScalingGroups"
            ],
            "Resource": "*"
        }
    ]
}
. Set Nomad Auto-Scaler to `enabled`

> . Set Max Node Count* - This will overwrite what is currently set as the max for you ASG. It is recommended to keep this value and what was set in your terraform as the same.

Reviewer: ```suggestion
. Set Max Node Count* - This will overwrite what is currently set as the max for you ASG. It is recommended to keep this value and what was set in your Terraform as the same.
```

. Set Min Node Count* - This will overwrite what is currently set as the max for you ASG. It is recommended to keep this value and what was set in your terraform as the same.
. Select cloud provider: `AWS EC2`
. Add the region of the auto scaling group

> . Add the nomad autoscaler user's access key

Reviewer: ```suggestion
. Add the Nomad Autoscaler user's access key
```

. Add the nomad autoscaler user's secret key
. Add the name of the autoscaling Group your nomad clients were created in

===== GCP
. Create a service account for Nomad Autoscaler
  * Our link:https://github.com/CircleCI-Public/server-terraform/tree/main/nomad-gcp[nomad module] creates a service acount and outputs a file with the keys if you set variable `nomad_auto_scaler = true`. You may reference the examples in the link for more details. If you've already created the clients, you can simply update the variable and run `terraform apply`. The created user's key will be available in a file named `nomad-as-key.json`.
  * You may also create a nomad gcp service account manually. The service account will need the role `compute.admin`.
. Set Nomad Auto-Scaler to `enabled`
. Set Maximum Node Count*
. Set Minimum Node Count*
. Select cloud provider: `Google Cloud Platform`
. Add your Project ID
. Add Managed Instance Group Name
. Instance group type: link:https://cloud.google.com/compute/docs/instance-groups/#types_of_managed_instance_groups[Zonal or Regional].
. JSON of GCP service account for Nomad Auto-Scaler

```

> * The maximum and minimum nomad client count will overwrite the corresponding values set when you created your autoscaling group or managed instance group. It is recommended that you keep these values and those used in your terraform the same so that the 2 don't compete.

Reviewer: ```suggestion
* The maximum and minimum nomad client count will overwrite the corresponding values set when you created your autoscaling group or managed instance group. It is recommended that you keep these values and those used in your Terraform the same so that the two don't compete.
```
